### ISSUE TITLE
[core.regexp] add matchEntirely and matchAsSuffix

### ISSUE THREAD

**Issue by cedvdb** [dart-lang/sdk#45586]
There is the method `matchAsPrefix` already. It seems weird the two other alternatives are not present: matchAsSuffix, matchEntirely.

A replace with template would be nice as well to replace groups.

```
RegExp('hello').matchAsPrefix('hello you');

// match entirely
// you don't really convey the intent here, especially if the pattern comes from a variable you 
// then have to check the string to be sure the pattern doesn't already includes the start and end codes. 
// you also cannot use a raw string with the char $ at the end making it even less lisible
RegExp('^hello$').firstMatch('hello');

final pattern = r'(hello)(.*)';
final template = r'bonjour $2';
// this code doesn't exists but would produce "bonjour you", swift has something like that, I believe java does too
RegExp(pattern).match('hello you').replaceWithTemplate(template);
```

**Comment by lrhn** [comment]
tl;dr: Can't be efficiently compiled to JavaScript.

The `matchAsPrefix` functionality isn't really there to only match at the beginning of the string. You could just match with `^(?:$orioginal)` instead. It's there because the `start` parameter allows you to try a match at any position, something `^` cannot do. Having it allows you to implement your own search for the next match, an anchored match at any position, rather than the default behavior which is to search for the next match automatically.

The reason we only have `matchAsPrefix` and not `matchAsSuffix` is that RegExps are traditionally matching *forward*. It's easy to say "Try to match from this position" because the underlying machinery does that all the time, but "try to find a match ending at this position" is not nearly as easy (not even with look-behinds). We need to be able to implement the functionality efficiently, also when compiling to JavaScript using their native RegExp. That means we cannot do anything which isn't supported at least somwwhat efficiently by that.

For `matchEntireString`, you can also do `^(?:$original)\$`. That does work, whether the original regexp contains anchors or not, as long as you avoid `multiline: true`.
To do something useful, it should work on a substring, something like `re.matchSubstring(source, start, end)` which would accept only a match starting at `start` and ending at `end`.
Again, the underlying RegExp machinery doesn't really allow us to implement this efficiently. (Creating the substring is not considered efficient, and would still might not work correctly if the regexp contains lookaheads).

**Comment by mraleph** [comment]
Given the @lrhn I will just close this.

**Comment by cedvdb** [comment]
> tl;dr: Can't be efficiently compiled to JavaScript.
> 
> The `matchAsPrefix` functionality isn't really there to only match at the beginning of the string. You could just match with `^(?:$orioginal)` instead. It's there because the `start` parameter allows you to try a match at any position, something `^` cannot do. Having it allows you to implement your own search for the next match, an anchored match at any position, rather than the default behavior which is to search for the next match automatically.
> 
> The reason we only have `matchAsPrefix` and not `matchAsSuffix` is that RegExps are traditionally matching _forward_. It's easy to say "Try to match from this position" because the underlying machinery does that all the time, but "try to find a match ending at this position" is not nearly as easy (not even with look-behinds). We need to be able to implement the functionality efficiently, also when compiling to JavaScript using their native RegExp. That means we cannot do anything which isn't supported at least somwwhat efficiently by that.
> 
> For `matchEntireString`, you can also do `^(?:$original)\$`. That does work, whether the original regexp contains anchors or not, as long as you avoid `multiline: true`.
> To do something useful, it should work on a substring, something like `re.matchSubstring(source, start, end)` which would accept only a match starting at `start` and ending at `end`.
> Again, the underlying RegExp machinery doesn't really allow us to implement this efficiently. (Creating the substring is not considered efficient, and would still might not work correctly if the regexp contains lookaheads).

It sounds like it is implemetable in javascript. If I understand correctly you can use matchAsPrefix and check the underlying lengths. Whether that's efficient or not idk.

for match entirely I have this:

```
extension MatchEntirely on RegExp {
  Match? matchEntirely(String pattern) {
    if (!pattern.endsWith(r'$')) {
      pattern = pattern + r'$';
    }
    return matchAsPrefix(pattern);
  }
}
```

**Comment by lrhn** [comment]
A `matchEntirely` on `RegExp` (not on `Pattern` in general) could definitely be created. Something like: 
```dart
Match? matchEntirely(String input) => RegExp("^(?:${this.pattern})\$").firstMatch(input);
```

My point is that because it's easy as this, it's not really needed as a feature. You can always write your pattern that way to begin with, and in the vast majority of use-cases, you have the pattern directly in your source code.
If the method cannot take a start/end position, like `matchAsPrefix` takes a start position, then it won't be worth having an extra method.
If it *does* have a start *and end*, then it's not efficiently implementable in JavaScript. Possibly not on the VM either, depending on the restrictions of the underlying regexp engine (it's an implementation of JavaScript regexps, so it doesn't have to support an end position different from the end of the string.

**Comment by cedvdb** [comment]
> then it won't be worth having an extra method.

Happy to disagree, for readability of code I think it would be worth it. Case closed then, I won't push it.
